src/Database/PostgreSQL/Consumers/Components.hs

```diff
@@ -169,35 +169,42 @@ spawnMonitor ConsumerConfig{..} cs cid = forkP "monitor" . forever $ do
       else do
         logInfo_ $ "Consumer is not registered"
         throwM ThreadKilled
-  -- Freeing jobs locked by inactive consumers needs to happen
-  -- exactly once, otherwise it's possible to free it twice, after
-  -- it was already marked as reserved by other consumer, so let's
-  -- run it in serializable transaction.
-  (inactiveConsumers, freedJobs) <- runDBT cs tsSerializable $ do
+  (inactiveConsumers, freedJobs) <- runDBT cs ts $ do
     now <- currentTime
-    -- Delete all inactive (assumed dead) consumers and get their ids.
-    runSQL_ $ smconcat [
-        "DELETE FROM" <+> raw ccConsumersTable
-      , "  WHERE last_activity +" <?> iminutes 1 <+> "<= " <?> now
-      , "    AND name =" <?> unRawSQL ccJobsTable
-      , "  RETURNING id::bigint"
+    -- Reserve all inactive (assumed dead) consumers and get their ids. We don't
+    -- delete them here, because if the coresponding reserved_by column in the
+    -- jobs table has an IMMEDIATE foreign key with the ON DELETE SET NULL
+    -- property, we will not be able to determine stuck jobs in the next step.
+    runSQL_ $ smconcat
+      [ "SELECT id::bigint"
+      , "FROM" <+> raw ccConsumersTable
+      , "WHERE last_activity +" <?> iminutes 1 <+> "<= " <?> now
+      , "  AND name =" <?> unRawSQL ccJobsTable
+      , "FOR UPDATE"
```

Comment:
same, just in case, no skip locked ?

Reply:
I added skip locked to both queries as it looks like the safest option.